### PR TITLE
feat(dapp-connector): Add CIP-30 support

### PR DESCRIPTION
**Pull request by michalszorad**
Resolves [YOMO-1191](https://emurgo.atlassian.net/jira/software/c/projects/YOMO/issues/YOMO-1191)

- [x] CIP30-getExtensions
- [x] CIP30-getNetworkId
- [x] CIP30-getUtxos
- [x] CIP30-getCollateral
- [x] CIP30-getBalance
- [x] CIP30-getUsedAddresses
- [x] CIP30-getUnusedAddresses
- [x] CIP30-getChangeAddress
- [x] CIP30-getRewardAddresses
- [x] CIP30-signTx
- [x] CIP30-submitTx
- [x] confirm connection modal
- [x] refactor (serialization, CSL, todos, utxos, translations, dapp-connector/src)
- [x] tests
- [x] revert project.pbxproj
- [x] add limit to 3 utxos in collateral

TX Flow will be finished in https://emurgo.atlassian.net/browse/YOMO-1198
HW support to be added in https://emurgo.atlassian.net/browse/YOMO-1442

[YOMO-1191]: https://emurgo.atlassian.net/browse/YOMO-1191?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by michalszorad** [comment]
Some tests  and the signData are still to be implemented, see more details here https://emurgo.slack.com/archives/GCN6JLZK7/p1715936664148259

File: apps/wallet-mobile/src/features/Discover/common/hooks.tsx [hunk 1]
```diff
@@ -52,3 +54,26 @@ const getInitScript = (sessionId: string, dappConnector: DappConnectorManager) =
...
+      closeModal()
+      return result
+    }
+
+    if (wallet.isHW) {
+      throw new Error('Not implemented yet')
+    }
+
+    if (wallet.isEasyConfirmationEnabled) {
+      openModal('Confirm TX', <ConfirmRawTxWithOs onConfirm={handleOnConfirm} />, undefined, onClose)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/common/hooks.tsx]: intl.

File: apps/wallet-mobile/src/features/Discover/common/hooks.tsx [hunk 2]
```diff
@@ -52,3 +54,26 @@ const getInitScript = (sessionId: string, dappConnector: DappConnectorManager) =
...
+      return
+    }
+
+    openModal('Confirm TX', <ConfirmRawTxWithPassword onConfirm={handleOnConfirm} />, undefined, onClose)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/common/hooks.tsx]: :arrow_up:

File: apps/wallet-mobile/src/features/Discover/useCases/BrowseDapp/BrowserTabBar.tsx [hunk 3]
```diff
@@ -56,7 +56,7 @@ export const BrowserTabBar = ({webViewRef, webViewState}: Props) => {
   }
 
   return (
-    <View style={[styles.root, styles.shadow, {paddingBottom: insets.bottom + 12}]}>
+    <View style={[styles.root, styles.shadow, {height: 46 + insets.bottom, paddingBottom: insets.bottom}]}>
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/useCases/BrowseDapp/BrowserTabBar.tsx]: won't it account along with NavTabBottom?

File: apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx [hunk 4]
```diff
@@ -33,6 +33,8 @@ export const DAppListItem = ({dApp, connected, onPress}: Props) => {
 
   const [isPressed, setIsPressed] = React.useState(false)
 
+  const logo = dApp.logo.length === 0 ? getDappFallbackLogo(dApp.origins[0]) : dApp.logo
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx]: what is origins[0], can it be destructed? (avoid index accessing)

File: apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx [hunk 5]
```diff
@@ -99,11 +99,9 @@ export const DAppListItem = ({dApp, connected, onPress}: Props) => {
       <View style={styles.dAppItemContainer}>
         {isGoogleSearchItem(dApp) ? (
           <Icon.Google />
-        ) : showIconPlaceholder ? (
-          <EmptyIcon />
         ) : (
           <View>
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx]: is this view needed?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts [hunk 6]
```diff
@@ -1465,6 +1463,47 @@ export class ByronWallet implements YoroiWallet {
       isEasyConfirmationEnabled: this.isEasyConfirmationEnabled,
     }
   }
+
+  CIP30getBalance(tokenId = '*'): Promise<CSL.Value> {
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts]: @michaeljscript how about having a `dappTranslator` object that receives a wallet (this) and returns the 30+95 (cardano interface) methods? (avoid starting with capital if not a constructor or component). e.g dapp = yadayada(this) -> accessing thought, wallet.dappTranslator.getBalance() ...

File: apps/wallet-mobile/src/yoroi-wallets/cardano/cip30.ts
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/cip30.ts:1]: @michaeljscript I think it should start wrapping the CSL to release memory all this manipulation pointer will sit on memory till closing the app. (there is an example in the lib and in the backend-zero.

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts [hunk 7]
```diff
@@ -975,6 +979,47 @@ export const makeShelleyWallet = (constants: typeof MAINNET | typeof TESTNET | t
       return doesCardanoAppVersionSupportCIP36(await getCardanoAppMajorVersion(this.hwDeviceInfo, useUSB))
     }
 
+    CIP30getBalance(tokenId = '*'): Promise<CSL.Value> {
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: :arrow_up:

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts [hunk 8]
```diff
@@ -13,13 +13,29 @@ export const collateralConfig: CollateralConfig = {
...
+}
+
 export function isAmountInCollateralRange(amount: RawUtxo['amount'], {maxLovelace, minLovelace}: CollateralConfig) {
   const value = new BigNumber(asQuantity(amount))
   const min = new BigNumber(minLovelace)
   const max = new BigNumber(maxLovelace)
   return value.gte(min) && value.lte(max)
 }
 
+export const findCollateralCandidates = (
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: shouldn't also receive a limit (3) of records to return ? when returning for CIP30?

File: packages/dapp-connector/package.json [hunk 9]
```diff
@@ -107,7 +107,7 @@
     "src/connector.js"
   ],
   "jest": {
-    "collectCoverage": true,
+    "collectCoverage": false,
```
Comment (stackchain) [inline packages/dapp-connector/package.json]: 👀

File: packages/dapp-connector/src/adapters/api.ts [hunk 10]
```diff
@@ -24,21 +24,43 @@ export const dappConnectorApiMaker = ({request}: {request: FetchData} = initialD
...
+        .map(modifyDapp),
       filters: value.filters,
     }
   }
   return {getDApps}
 }
 
+const modifyDapp = (dapp: DappResponse): DappResponse => {
+  if (dapp.id === 'wingriders') {
+    return {...dapp, origins: ['https://www.wingriders.com', 'https://app.wingriders.com']}
```
Comment (stackchain) [inline packages/dapp-connector/src/adapters/api.ts]: any reason to don't be an object? {site, app} ?

File: packages/dapp-connector/src/resolver.ts [hunk 11]
```diff
@@ -131,24 +237,40 @@ const handleMethod = async (
...
-    const result = await handleMethod(method, params, {origin: trustedOrigin, wallet, storage})
+    const result = await handleMethod(method, params, {origin: trustedOrigin, wallet, storage, supportedExtensions})
     if (method !== LOG_MESSAGE_EVENT) sendMessage(id, result)
   } catch (error) {
     sendMessage(id, null, error as Error)
   }
 }
 
-type Wallet = {
+export type ResolverWallet = {
```
Comment (stackchain) [inline packages/dapp-connector/src/resolver.ts:256]: temporary I'm guessing until Wallet -> @yoroi/types ? prefer the new `network`.

**Review by stackchain — APPROVED** [review]
looking pretty solid

**Review by jorbuedo — APPROVED** [review]

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts (hunk 6 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts]: Yes, sounds good

File: packages/dapp-connector/src/adapters/api.ts (hunk 10 above)
Comment (michalszorad) [reply packages/dapp-connector/src/adapters/api.ts]: Sorry, this will be moved to the JSON

**Comment by michalszorad** [comment]
The `signData` may need changes in an external library and will be done in a different PR. See YOMO-1447

File: packages/dapp-connector/package.json (hunk 9 above)
Comment (michalszorad) [reply packages/dapp-connector/package.json]: Reverted. I've turned it off for testing.

File: packages/dapp-connector/src/adapters/api.ts (hunk 10 above)
Comment (stackchain) [reply packages/dapp-connector/src/adapters/api.ts]: 0 worries, IMHO, this object should come from the response already, so this mapping is not needed in the client.

Comment (michalszorad) [reply packages/dapp-connector/src/adapters/api.ts]: It was mocked for testing purposes, the data is now updated and coming from https://daehx1qv45z7c.cloudfront.net/data.json

File: apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx (hunk 4 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx]: Refactored not to use origins

File: apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx (hunk 5 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Discover/useCases/SelectDappFromList/DAppListItem/DAppListItem.tsx]: Not needed

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts (hunk 8 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: Is that a limit of 3 UTXOS? Hmm, I could not find any limit anywhere in the documentation.

Comment (stackchain) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: yap, please ping @vsubhuman

File: apps/wallet-mobile/src/yoroi-wallets/cardano/cip30.ts
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/cip30.ts:1]: I've added a modified version of the wrapped-csl from backend. However I've realised the CIP30 helpers return CSL data, so in some cases we can't release the memory. A solution could be if the helpers returned already serialised data.

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts (hunk 6 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts]: I've removed the `CIP30` methods from YoroiWallet interface and created a `cip30ExtensionMaker` which takes wallet and returns cip30 methods

File: apps/wallet-mobile/src/features/Discover/useCases/BrowseDapp/BrowserTabBar.tsx (hunk 3 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Discover/useCases/BrowseDapp/BrowserTabBar.tsx]: Refactored to have fixed height, and insets are computed in the parent screen

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts (hunk 7 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: Updated

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts (hunk 8 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: I'll be adding a limit of 5 utxos as discussed on Slack

Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: I'll be changing this to 3

Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utxoManager/utxos.ts:29]: That's now updated

**Comment by michalszorad** [comment]
@stackchain, that should be now all updated. Am I good to merge?

**Review by stackchain — APPROVED** [review]

**Pull request merged**